src/library/coverartdelegate.cpp

```diff
@@ -108,16 +107,16 @@ void CoverArtDelegate::paintItem(QPainter *painter,
     info.hash = index.sibling(index.row(), m_iCoverHashColumn).data().toUInt();
     info.trackLocation = index.sibling(index.row(), m_iTrackLocationColumn).data().toString();
 
+    double scaleFactor = getDevicePixelRatioF(dynamic_cast<QWidget*>(parent()));
     // We listen for updates via slotCoverFound above and signal to
     // BaseSqlTableModel when a row's cover is ready.
     QPixmap pixmap = pCache->requestCover(info, this, option.rect.width(),
-                                          m_bOnlyCachedCover, true);
+                                          scaleFactor, m_bOnlyCachedCover, true);
     if (!pixmap.isNull()) {
-        int width = math_min(pixmap.width(), option.rect.width());
-        int height = math_min(pixmap.height(), option.rect.height());
-        QRect target(option.rect.x(), option.rect.y(),
-                     width, height);
-        QRect source(0, 0, target.width(), target.height());
+        int width = math_min(pixmap.width(), option.rect.width()) * scaleFactor;
+        int height = math_min(pixmap.height(), option.rect.height()) * scaleFactor;
+        QRect target(option.rect.x(), option.rect.y(), width, height);
```

Comment:
Why do we have to scale manually to the device resolution instead of letting Qt do the work by consistently using setDevicePixelRatio() and doing all sizing in device independent coordinates? This is how I thought it is supposed to work.

Reply:
Hmm, I'll take another look at this logic. pixmap.width()/height() should be in scaled sizes already, but it looks like math_min picks the option.rect.width()/height(), which then needs to be scaled.

Reply:
When I was first writing this, I multiplied by scaleFactor everywhere until I got it to work without really understanding what was going on. I may be able to simplify the code now...

Reply:
All this manual scaling is confusing, I still don't get it. Why doesn't Qt provide a consistent measurement system? My assumption was that setting *device pixel ratio* does all the magic for us. These  conversions back and forth between different coordinate systems are really error prone.

Reply:
I agree, it is awfully confusing.

Reply:
I think the original code is now outdated since we have non ^2 scale favtors.
Originally this code prevents us from upscaling an actual 15×15 to 20x20 for instance. This will likely look blurry. Instead we print the 15x15 and pad it for 20x20

This approach still makes sense, but we need to considder the scaleFactor.
If it is 1 or 2 the code works. If it is 1.5 the code fails.
20x20 is in this case 30x30 and the pixmap will look perfectly scaled by 2.

I must admit that this is a constructed example, but it is relevant.